packages/router/src/grpc/offscreen-client.ts

```diff
@@ -0,0 +1,47 @@
+import { WitnessAndBuildRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
+import { WitnessData } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import { InternalRequest, InternalResponse } from '@penumbra-zone/types/src/internal-msg/shared';
+import {
+  ActionBuildMessage,
+  OffscreenMessage,
+} from '@penumbra-zone/types/src/internal-msg/offscreen-types';
+
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+
+export const offscreenClient = {
+  buildAction: (
+    arg: WitnessAndBuildRequest,
+    witness: WitnessData,
+    fullViewingKey: string,
+  ): Promise<InternalResponse<ActionBuildMessage>> =>
+    sendOffscreenMessage<ActionBuildMessage>({
+      type: 'BUILD_ACTION',
+      request: {
+        transactionPlan: arg.transactionPlan!.toJson(),
+        witness: witness.toJson(),
+        fullViewingKey,
+        length: arg.transactionPlan!.actions.length,
+      },
+    }),
+};
+
+export const sendOffscreenMessage = async <T extends OffscreenMessage>(
+  req: InternalRequest<T>,
+): Promise<InternalResponse<ActionBuildMessage>> => {
+  await chrome.offscreen.createDocument({
+    url: OFFSCREEN_DOCUMENT_PATH,
+    reasons: [chrome.offscreen.Reason.WORKERS],
+    justification: 'spawn web workers from offscreen document',
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+  const result = (await chrome.runtime.sendMessage({
+    ...req,
+  })) as InternalResponse<ActionBuildMessage>;
+  if ('error' in result) throw new Error('failed to build action');
+
+  // Close offscreen document
+  await chrome.offscreen.closeDocument();
```

Comment:
we should probably not be killing the offscreen when it's done - multiple callers could be talking to the same `offscreen.html`. there won't be crosstalk in the responses, because direct responses are awaited. but it *is* possible that one handler may kill an offscreen doc that another handler is currently pending.

leaving the doc alive should be ok for now. it maintains one message listener, and kills workers when it's done with them, so it should be very lightweight. let's remove this line and i'll make an issue on it

```suggestion
```

Reply:
Is it good practice to generally keep the offscreen API document open? I'm not sure

Reply:
Actually, seems like we need to close the document, no? Initiating a new transaction throws `"Only a single offscreen document may be created"`

Reply:
it isn't! but offscreen may simultaneously perform work for multiple callers. it's unclear how to identify that we no longer need the offscreen document.

one solution would be to make offscreen responsible for its own termination, but this seems incorrect.

maybe attach a counter to offscreenClient that is incremented/decremented when handlers launch and resolve, and kill when it's zero. there is still a race condition there, but it will prevent killing other callers' work in most cases.

Reply:
Yeah, I feel like we can just leave it open. When the service worker spins down, it should kill the offscreen too (I think?).

Reply:
we can't keep it open. The user won't be able to send another transaction directly after. 

> Actually, seems like we need to close the document, no? Initiating a new transaction throws `"Only a single offscreen document may be created"`

If we keep the offscreen document open, this is still the case. The user can't issue a transaction directly after the first one.

Reply:
> it isn't! but offscreen may simultaneously perform work for multiple callers. it's unclear how to identify that we no longer need the offscreen document.

@turbocrime I don't see in what scenario there would be multiple callers (at least for now). Let's keep `await chrome.offscreen.closeDocument();` for now since it's good practice, and we can address this in a separate issue.